### PR TITLE
Add Counter Example

### PR DESCRIPTION
**Pull request by patjackson52**
- following structure of JS redux.  Examples under example directory
 - basic counter example https://redux.js.org/introduction/examples
 - android only for now
 - bumped kotlin to 1.3.50 & redux version to 0.2.7

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -6,7 +6,7 @@ version: 2.1
 jobs:
   publish_release:
     docker:
-      - image: circleci/android:api-29
+      - image: circleci/openjdk:8-jdk
```
Comment (AOrobator) [inline .circleci/config.yml]: Why the change to a more generic docker image?

File: build.gradle [hunk 2]
```diff
@@ -12,6 +12,7 @@ buildscript {
     dependencies {
         classpath deps.plugins.kotlin
         classpath deps.plugins.dokka
+        classpath deps.plugins.android
```
Comment (AOrobator) [inline build.gradle]: I know this is out of scope for this PR, but what do you think about moving over to the Kotlin Gradle DSL?

File: examples/counter/android/build.gradle [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        release {
+            minifyEnabled false
+            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
+        }
+        debug {
+            // MPP libraries don't currently get this resolution automatically
+            matchingFallbacks = ['release']
+        }
+    }
+//    signingConfigs {
```
Comment (AOrobator) [inline examples/counter/android/build.gradle]: Let's avoid committing dead code

File: examples/counter/android/build.gradle [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+//    }
+    packagingOptions {
+        exclude 'META-INF/*.kotlin_module'
+    }
+    tasks.lint.enabled = false
+}
+
+
+dependencies {
+    implementation fileTree(dir: 'libs', include: ['*.jar'])
```
Comment (AOrobator) [inline examples/counter/android/build.gradle]: This line can probably be deleted, unless you're actually including jars

File: examples/counter/android/build.gradle [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    implementation"org.jetbrains.kotlin:kotlin-stdlib-jdk7:$project.versions.kotlin"
```
Comment (AOrobator) [inline examples/counter/android/build.gradle]: We should probs use Kotlin JDK 8 if we're compiling with Java 8

File: examples/counter/android/build.gradle [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    implementation 'androidx.appcompat:appcompat:1.1.0-rc01'
+
+    implementation project(':examples:counter:common')
+    implementation project(':lib')
```
Comment (AOrobator) [inline examples/counter/android/build.gradle]: lib should be renamed. it's very generic. How about redux-kotlin-core?

File: examples/counter/android/src/main/java/org/reduxkotlin/example/counter/MainActivity.kt [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+import kotlinx.android.synthetic.main.activity_main.*
+import org.reduxkotlin.StoreSubscription
+import org.reduxkotlin.createStore
+import org.reduxkotlin.examples.counter.Decrement
+import org.reduxkotlin.examples.counter.Increment
+import org.reduxkotlin.examples.counter.reducer
+
+/**
+ * This is a sample of basic redux behavior.
+ * This is NOT best practice for structuring a multiplatform App.
```
Comment (AOrobator) [inline examples/counter/android/src/main/java/org/reduxkotlin/example/counter/MainActivity.kt]: Is there a plan to show best practices?

File: examples/counter/android/src/main/java/org/reduxkotlin/example/counter/MainActivity.kt [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        btn_async.setOnClickListener { incrementAsync() }
+        btn_increment_if_odd.setOnClickListener { incrementIfOdd() }
+    }
+
+    private fun render(state: Int) {
+        txt_label.text = "Clicked: $state times"
+    }
+
+    private fun incrementIfOdd() {
+        if (store.state % 2 != 0) {
```
Comment (AOrobator) [inline examples/counter/android/src/main/java/org/reduxkotlin/example/counter/MainActivity.kt]: Should this kind of logic go in the reducer?

File: examples/counter/android/src/main/res/drawable/ic_arrow_downward_white_24dp.xml [hunk 9]
```diff
@@ -0,0 +1,5 @@
+<vector android:height="24dp" android:tint="#FFFFFF"
```
Comment (AOrobator) [inline examples/counter/android/src/main/res/drawable/ic_arrow_downward_white_24dp.xml]: This should be formatted so everything is on a separate line.

File: examples/counter/android/src/main/res/drawable/ic_arrow_upward_black_24dp.xml [hunk 10]
```diff
@@ -0,0 +1,5 @@
+<vector android:height="24dp" android:tint="#FFFFFF"
```
Comment (AOrobator) [inline examples/counter/android/src/main/res/drawable/ic_arrow_upward_black_24dp.xml]: Formatting

File: examples/counter/android/src/main/res/layout/activity_main.xml [hunk 11]
```diff
@@ -0,0 +1,49 @@
+<?xml version="1.0" encoding="utf-8"?>
+
```
Comment (AOrobator) [inline examples/counter/android/src/main/res/layout/activity_main.xml]: We should get rid of this new line

File: examples/counter/android/src/main/res/layout/activity_main.xml [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (AOrobator) [inline examples/counter/android/src/main/res/layout/activity_main.xml]: Curious on what your thoughts are on using ConstraintLayout. I understand it's not really the point of this app, but it'd be nice to use best practices in general

File: examples/counter/android/src/main/res/values/strings.xml [hunk 13]
```diff
@@ -0,0 +1,5 @@
+<resources>
+    <string name="app_name">Counter Example</string>
+
```
Comment (AOrobator) [inline examples/counter/android/src/main/res/values/strings.xml]: Lots of empty lines we can delete

File: examples/counter/android/src/main/res/values/styles.xml [hunk 14]
```diff
@@ -0,0 +1,6 @@
+<resources xmlns:tools="http://schemas.android.com/tools">
+
+    <style name="CounterAppTheme" parent="Theme.AppCompat.DayNight.NoActionBar">
```
Comment (AOrobator) [inline examples/counter/android/src/main/res/values/styles.xml]: This style has an empty body

```suggestion
    <style name="CounterAppTheme" parent="Theme.AppCompat.DayNight.NoActionBar"/>
```

File: examples/counter/common/build.gradle [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+            kotlin.srcDir('src/jvmMain/kotlin')
+            dependencies {
+                implementation kotlin("stdlib")
+            }
+        }
+        jvmTest {
+            dependencies {
+                implementation kotlin("test")
+                implementation kotlin("test-junit")
+                implementation "org.spekframework.spek2:spek-dsl-jvm:$project.versions.spek"
```
Comment (AOrobator) [inline examples/counter/common/build.gradle]: I've had a good amount of success with [buildSrcVersions](https://github.com/jmfayard/buildSrcVersions) for generating code to manage dependencies across modules. Will also give you type completion in the ide, once the buildSrc module is set up

File: examples/counter/common/build.gradle [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+        }
+        jsMain {
+            kotlin.srcDir('src/jsMain/kotlin')
+            dependencies {
+                implementation kotlin("stdlib-js")
+            }
+            compileKotlinJs {
+                kotlinOptions.metaInfo = true
+                kotlinOptions.sourceMap = true
+                kotlinOptions.suppressWarnings = true
```
Comment (AOrobator) [inline examples/counter/common/build.gradle]: Are you actually getting warnings now that you'd like to ignore? Otherwise it's probably best to be strict and keep the project clean by treating warnings as errors

File: examples/counter/common/src/commonMain/kotlin/org/reduxkotlin/examples/counter/Reducer.kt [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+
+import org.reduxkotlin.Reducer
+
+
+class Increment
+class Decrement
+
+val reducer: Reducer<Int> = {state, action ->
+    when (action) {
+        is Increment -> state + 1
```
Comment (AOrobator) [inline examples/counter/common/src/commonMain/kotlin/org/reduxkotlin/examples/counter/Reducer.kt]: A sealed reducer could eliminate the else statement

File: examples/counter/common/src/commonMain/kotlin/org/reduxkotlin/examples/counter/Reducer.kt [hunk 18]
```diff
@@ -0,0 +1,15 @@
+package org.reduxkotlin.examples.counter
+
+import org.reduxkotlin.Reducer
+
+
+class Increment
+class Decrement
+
+val reducer: Reducer<Int> = {state, action ->
```
Comment (AOrobator) [inline examples/counter/common/src/commonMain/kotlin/org/reduxkotlin/examples/counter/Reducer.kt]:
```suggestion
val reducer: Reducer<Int> = { state, action ->
```

File: examples/counter/common/src/test/kotlin/org/reduxkotlin/examples/counter/ReducerTest.kt [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+
+import ch.tutteli.atrium.api.cc.en_GB.toBe
+import org.spekframework.spek2.Spek
+import org.spekframework.spek2.style.specification.describe
+import ch.tutteli.atrium.verbs.expect
+
+
+object CounterSpek : Spek({
+
+    describe("reducers") {
```
Comment (AOrobator) [inline examples/counter/common/src/test/kotlin/org/reduxkotlin/examples/counter/ReducerTest.kt]: Why is this describe nested? Seems like there should only be one, not two

File: lib/src/commonMain/kotlin/org/reduxkotlin/Definitions.kt [hunk 20]
```diff
@@ -6,6 +6,12 @@ package org.reduxkotlin
  */
 typealias Reducer<State> = (state: State, action: Any) -> State
 
+/**
+ * Reducer for a particular subclass of actions.  Useful for Sealed classes &
+ * exhaustive when statements.  see [reducerForActionType]
```
Comment (AOrobator) [inline lib/src/commonMain/kotlin/org/reduxkotlin/Definitions.kt]:
```suggestion
 * exhaustive when statements.  See [reducerForActionType].
```

File: lib/src/commonMain/kotlin/org/reduxkotlin/Definitions.kt [hunk 21]
```diff
@@ -52,3 +58,26 @@ fun <State> middleware(dispatch: (Store<State>, next: Dispatcher, action: Any) -
...
+
+
+/**
+ * Convenience function for creating a [ReducerForActionType]
+ * usage:
+ *   sealed class LoginScreenAction
+ *   data class LoginComplete(val user: User): LoginScreenAction()
+ *
+ *   val loginReducer = sealedReducer<AppState, LoginAction> { state, action ->
+ *       when(action) {
```
Comment (AOrobator) [inline lib/src/commonMain/kotlin/org/reduxkotlin/Definitions.kt]: Might be worth showing how 2 sealedReducers can be combined and passed to a Store

File: lib/src/commonMain/kotlin/org/reduxkotlin/Definitions.kt [hunk 22]
```diff
@@ -52,3 +58,26 @@ fun <State> middleware(dispatch: (Store<State>, next: Dispatcher, action: Any) -
...
+ *           is LoginComplete -> state.copy(user = action.user)
+ *       }
+ *   }
+ */
+inline fun <TState, reified TAction> reducerForActionType(crossinline reducer: ReducerForActionType<TState, TAction>): Reducer<TState> {
+    return {state, action ->
```
Comment (AOrobator) [inline lib/src/commonMain/kotlin/org/reduxkotlin/Definitions.kt]: have you considered adding detekt/static analysis for CI builds?

https://github.com/arturbosch/detekt

```suggestion
    return { state, action ->
```

File: lib/src/commonMain/kotlin/org/reduxkotlin/Definitions.kt [hunk 23]
```diff
@@ -52,3 +58,26 @@ fun <State> middleware(dispatch: (Store<State>, next: Dispatcher, action: Any) -
...
+        if (action is TAction) {
```
Comment (AOrobator) [inline lib/src/commonMain/kotlin/org/reduxkotlin/Definitions.kt]: This could be more concise if a when statement was used

**Comment by AOrobator** [comment]
This example should be added to the readme

File: .circleci/config.yml (hunk 1 above)
Comment (patjackson52) [reply .circleci/config.yml]: originally it was openjdk.  I ran into an issue with the installed gradle version not being compatible with the android plugin.  I tried using the android image as a solution, but have decided against it.  It didn't have a different installed gradle version.  I have switched to using the gradle wrapper as a fix.  If we start building apks or doing android tests we may need to change, but also need to consider iOS, so think that is part of a bigger solution.

File: build.gradle (hunk 2 above)
Comment (patjackson52) [reply build.gradle]: I like the idea of it.  I haven't used it much.  The difficulty I have ran into is nearly all examples online, or existing build.gradle files are groovy, and it is not always clear how to migrate to kotlin.  Have you used it much?  I would like to move to it eventually

File: examples/counter/android/build.gradle (hunk 3 above)
Comment (patjackson52) [reply examples/counter/android/build.gradle]: removed

File: examples/counter/android/src/main/java/org/reduxkotlin/example/counter/MainActivity.kt (hunk 7 above)
Comment (patjackson52) [reply examples/counter/android/src/main/java/org/reduxkotlin/example/counter/MainActivity.kt]: Yes, through other examples and documentation on the website.  This is strictly a port of the redux counter example.  I wanted to get something up as a template for other examples.  These will also let us test/exercise the core redux code within the repo.

File: examples/counter/android/src/main/java/org/reduxkotlin/example/counter/MainActivity.kt (hunk 8 above)
Comment (patjackson52) [reply examples/counter/android/src/main/java/org/reduxkotlin/example/counter/MainActivity.kt]: again not best practice here, just going based on JS example.  In a MP project, def don't do this.  it is putting logic into the platform/view.  This is one pattern of creating an action.   This is also a really contrived example.  One way of doing this would be have a separate action named `class IncrementIfOdd` and do the logic in the reducer.  keep in mind the reducer can not dispatch actions.

File: examples/counter/android/src/main/res/drawable/ic_arrow_downward_white_24dp.xml (hunk 9 above)
Comment (patjackson52) [reply examples/counter/android/src/main/res/drawable/ic_arrow_downward_white_24dp.xml]: done.  was just the way AS imports vector drawables

File: examples/counter/android/src/main/res/layout/activity_main.xml (hunk 12 above)
Comment (patjackson52) [reply examples/counter/android/src/main/res/layout/activity_main.xml]: Love constraintlayout and use it for almost everything, just was a bit quicker for me to do it this way.  Can we get this merged and do another PR for constraintlayout?

File: examples/counter/android/src/main/res/values/styles.xml (hunk 14 above)
Comment (patjackson52) [reply examples/counter/android/src/main/res/values/styles.xml]: done in my commit

File: examples/counter/common/build.gradle (hunk 15 above)
Comment (patjackson52) [reply examples/counter/common/build.gradle]: I've been looking at that and looks nice.  Want to make a ticket for updating on trello? - either to buildSrcVersions, or at least a standard across reduxkotlin, reduxkotlin-thunk, etc.   The other repos use a slightly different strategy

File: examples/counter/common/src/commonMain/kotlin/org/reduxkotlin/examples/counter/Reducer.kt (hunk 17 above)
Comment (patjackson52) [reply examples/counter/common/src/commonMain/kotlin/org/reduxkotlin/examples/counter/Reducer.kt]: True..I'd like to keep this a minimal example.  I see sealed classes as an optional, more advanced usage.  I would not point people to this for a "best practices" example

File: examples/counter/common/src/test/kotlin/org/reduxkotlin/examples/counter/ReducerTest.kt (hunk 19 above)
Comment (patjackson52) [reply examples/counter/common/src/test/kotlin/org/reduxkotlin/examples/counter/ReducerTest.kt]: hmm, following the js example again here.  I believe its nested here so you could have multiple reducers tested here, and nested under the reducers group.

File: lib/src/commonMain/kotlin/org/reduxkotlin/Definitions.kt (hunk 20 above)
Comment (patjackson52) [reply lib/src/commonMain/kotlin/org/reduxkotlin/Definitions.kt]: applyed in my commit

File: lib/src/commonMain/kotlin/org/reduxkotlin/Definitions.kt (hunk 21 above)
Comment (patjackson52) [reply lib/src/commonMain/kotlin/org/reduxkotlin/Definitions.kt]: added

File: lib/src/commonMain/kotlin/org/reduxkotlin/Definitions.kt (hunk 23 above)
Comment (patjackson52) [reply lib/src/commonMain/kotlin/org/reduxkotlin/Definitions.kt]: adjusted and formatting changed to shorter line breaks

File: examples/counter/common/build.gradle (hunk 16 above)
Comment (patjackson52) [reply examples/counter/common/build.gradle]: Good idea.  have not got into the JS side much yet.  Will remove.  Yes, I agree with treating warnings as errors.

File: build.gradle (hunk 2 above)
Comment (AOrobator) [reply build.gradle]: Yeah I’m using it fully in an Android project. I could make a PR, but no promises on the timeline

File: examples/counter/common/build.gradle (hunk 15 above)
Comment (AOrobator) [reply examples/counter/common/build.gradle]: Done
